File: 2023/7/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split(" "));

const cardStrengths = {
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  T: 10,
  J: 11,
  Q: 12,
  K: 13,
  A: 14,
};

const cardStrengths2 = {
  J: 1,
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  T: 10,
  Q: 12,
  K: 13,
  A: 14,
};

class Hand {
  constructor(hand, bid) {
    this.hand = hand;
    this.handSorted = this.sortHand(hand);
    this.bid = Number(bid);
    this.type = this.evalHand(this.handSorted);
    this.type2 = this.evalHand2(this.handSorted);
  }

  sortHand(hand) {
    return hand
      .split("")
      .sort((a, b) => cardStrengths[b] - cardStrengths[a])
      .join("");
  }

  evalHand(hand) {
    if (/(\w)\1{4}/.test(hand)) return 7; // Five of a kind
    if (/(\w)\1{3}/.test(hand)) return 6; // Four of a kind
    if (/(\w)\1{2}(\w)\2|(\w)\3(\w)\4{2}/.test(hand)) return 5; // Full house
    if (/(\w)\1{2}/.test(hand)) return 4; // Three of a kind
    if (/(\w)\1.*?(\w)\2/.test(hand)) return 3; // Two pair
    if (/(\w)\1/.test(hand)) return 2; // One pair
    return 1; // High card
  }

  evalHand2(hand) {
    let hand2 = hand.replace(/J/g, "");
    if (hand2.length === 5) return this.type; // No jokers
    const type = this.evalHand(hand2);
    if (hand2.length === 4) {
      // One joker
      if (type === 6) return 7; // Four of a kind => Five of a kind
      if (type === 4) return 6; // Three of a kind => Four of a kind
      if (type === 3) return 5; // Two pair => Full house
      if (type === 2) return 4; // One pair => Three of a kind
      return 2; // High card => One pair
    }
    if (hand2.length === 3) {
      // Two jokers
      if (type === 4) return 7; // Three of a kind => Five of a kind
      if (type === 2) return 6; // One pair => Four of a kind
      return 4; // High card => Three of a kind
    }
    if (hand2.length === 2) {
      // Three jokers
      if (type === 2) return 7; // One pair => Five of a kind
      return 6; // High card => Four of a kind
    }
    return 7; // Four or all jokers => Five of a kind
  }
}

const hands = input.map(([a, b]) => new Hand(a, b));

hands.sort((a, b) => {
  if (a.type !== b.type) return a.type - b.type;
  for (let i = 0; i < 5; i++) {
    if (a.hand[i] !== b.hand[i])
      return cardStrengths[a.hand[i]] - cardStrengths[b.hand[i]];
  }
  return 0;
});

const part1 = hands.reduce((a, b, i) => a + b.bid * (i + 1), 0);
console.log(part1);

hands.sort((a, b) => {
  if (a.type2 !== b.type2) return a.type2 - b.type2;
  for (let i = 0; i < 5; i++) {
    if (a.hand[i] !== b.hand[i])
      return cardStrengths2[a.hand[i]] - cardStrengths2[b.hand[i]];
  }
  return 0;
});

const part2 = hands.reduce((a, b, i) => a + b.bid * (i + 1), 0);
console.log(part2);

File: 2023/7/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	hands := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(hands))
	fmt.Println(part2(hands))
}

var cardStrengthMap = makeCardStrengthMap("23456789TJQKA")
var cardStrengthMap2 = makeCardStrengthMap("J23456789TQKA")

func makeCardStrengthMap(cardsInOrder string) map[byte]int {
	res := make(map[byte]int, len(cardsInOrder))
	for i := range cardsInOrder {
		res[cardsInOrder[i]] = i
	}
	return res
}

type Hand struct {
	cards     string
	bid       int
	strength  int
	strength2 int
}

func NewHand(line string) *Hand {
	parts := strings.Split(line, " ")
	bid, err := strconv.Atoi(parts[1])
	if err != nil {
		panic(err)
	}
	return &Hand{parts[0], bid, calcHandStrength(parts[0], false), calcHandStrength(parts[0], true)}
}

func (h1 *Hand) play(h2 *Hand) bool {
	if h1.strength != h2.strength {
		return h1.strength < h2.strength
	}
	for i := range h1.cards {
		if h1.cards[i] != h2.cards[i] {
			return cardStrengthMap[h1.cards[i]] < cardStrengthMap[h2.cards[i]]
		}
	}
	return false
}

func (h1 *Hand) play2(h2 *Hand) bool {
	if h1.strength2 != h2.strength2 {
		return h1.strength2 < h2.strength2
	}
	for i := range h1.cards {
		if h1.cards[i] != h2.cards[i] {
			return cardStrengthMap2[h1.cards[i]] < cardStrengthMap2[h2.cards[i]]
		}
	}
	return false
}

func calcHandStrength(cards string, withJokers bool) int {
	cardMap := make(map[rune]int, len(cards))
	for _, card := range cards {
		cardMap[card]++
	}
	if withJokers {
		nJokers := cardMap['J']
		if nJokers == 5 {
			return 6
		}
		delete(cardMap, 'J')
		nMax := 0
		var cardToPretend rune
		for card, n := range cardMap {
			if n > nMax {
				nMax = n
				cardToPretend = card
			}
		}
		cardMap[cardToPretend] += nJokers
	}
	if len(cardMap) == 1 {
		// five of a kind
		return 6
	}
	if len(cardMap) == 5 {
		// high card
		return 0
	}
	if len(cardMap) == 4 {
		// one pair
		return 1
	}
	if len(cardMap) == 2 {
		for _, n := range cardMap {
			if n == 4 || n == 1 {
				// four a kind
				return 5
			}
			// full house
			return 4
		}
	}
	for _, n := range cardMap {
		if n == 3 {
			// three of a kind
			return 3
		} else if n == 2 {
			// return two pair
			return 2
		}
	}
	return -1
}

func parseInput(data []string) []*Hand {
	hands := make([]*Hand, len(data))
	for i, line := range data {
		hands[i] = NewHand(line)
	}
	return hands
}

func part1(hands []*Hand) int {
	sort.SliceStable(hands, func(i, j int) bool {
		return hands[i].play(hands[j])
	})
	res := 0
	for i, hand := range hands {
		res += (i + 1) * hand.bid
	}
	return res
}

func part2(hands []*Hand) int {
	sort.SliceStable(hands, func(i, j int) bool {
		return hands[i].play2(hands[j])
	})
	res := 0
	for i, hand := range hands {
		res += (i + 1) * hand.bid
	}
	return res
}
